File: scripts/reconstruct-b52-d12-8-motion-disocclusion.py
#!/usr/bin/env python3
"""Scalar Python structural-validity and adaptive-risk consumer for B52-D12.8."""
from __future__ import annotations

import argparse
import hashlib
import json
import math
import os
import sys
from pathlib import Path

import numpy as np


SPEC_SHA256 = "67722b1c8fafa0b83518e6e467de1adb9ca88bd32b7145f15be2d5627767b4d4"
INPUTS = {
    "previousRgba": ("previous.rgba32", 4),
    "currentRgba": ("current.rgba32", 4),
    "previousDepth": ("previous-depth.f32", 1),
    "currentDepth": ("current-depth.f32", 1),
    "previousOwner": ("previous-owner.f32", 1),
    "currentOwner": ("current-owner.f32", 1),
    "vector": ("vector.xy32", 2),
    "vectorNext": ("vector-next.xy32", 2),
}
OUTPUTS = {
    "adaptiveReconstructed": ("adaptive-reconstructed.rgba32", "<f4"),
    "reason": ("reason.u8", "u1"),
    "analyticOwner": ("analytic-owner.u8", "u1"),
    "structuralValid": ("structural-valid.u8", "u1"),
    "radius2Interior": ("radius2-interior.u8", "u1"),
    "radius3Interior": ("radius3-interior.u8", "u1"),
    "adaptiveInterior": ("adaptive-interior.u8", "u1"),
    "adaptiveRejected": ("adaptive-rejected.u8", "u1"),
    "riskRgb": ("risk.rgb64", "<f8"),
}
REASONS = {
    "UNREGISTERED": 0,
    "INVALID_CURRENT_ORACLE": 1,
    "INVALID_BOUNDS": 2,
    "INVALID_OWNER": 3,
    "INVALID_ALPHA": 4,
    "INVALID_DEPTH": 5,
    "VALID": 6,
}


def sha_bytes(value: bytes) -> str:
    return hashlib.sha256(value).hexdigest()


def sha_file(file_path: Path) -> str:
    digest = hashlib.sha256()
    with file_path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def canonical_hash(value: object) -> str:
    encoded = json.dumps(value, sort_keys=True, separators=(",", ":"), allow_nan=False).encode()
    return hashlib.sha256(encoded).hexdigest()


def arguments() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--spec", type=Path, required=True)
    parser.add_argument("--fixture", required=True)
    parser.add_argument("--repeat", type=int, choices=(1, 2), required=True)
    parser.add_argument("--input-dir", type=Path, required=True)
    parser.add_argument("--adapter-report", type=Path, required=True)
    parser.add_argument("--output-dir", type=Path, required=True)
    parser.add_argument("--report", type=Path, required=True)
    return parser.parse_args()


def rotation_xyz(values: list[float]) -> tuple[tuple[float, float, float], ...]:
    x, y, z = (float(value) for value in values)
    cx, sx = math.cos(x), math.sin(x)
    cy, sy = math.cos(y), math.sin(y)
    cz, sz = math.cos(z), math.sin(z)
    return (
        (cz * cy, cz * sy * sx - sz * cx, cz * sy * cx + sz * sx),
        (sz * cy, sz * sy * sx + cz * cx, sz * sy * cx - cz * sx),
        (-sy, cy * sx, cy * cx),
    )


def transform(row: dict) -> tuple[tuple[float, float, float], tuple[tuple[float, float, float], ...]]:
    return tuple(float(value) for value in row["location"]), rotation_xyz(row["rotationEuler"])


def add(left, right):
    return (left[0] + right[0], left[1] + right[1], left[2] + right[2])


def subtract(left, right):
    return (left[0] - right[0], left[1] - right[1], left[2] - right[2])


def scale(vector, value: float):
    return (vector[0] * value, vector[1] * value, vector[2] * value)


def dot(left, right) -> float:
    return left[0] * right[0] + left[1] * right[1] + left[2] * right[2]


def mat_vec(matrix, vector):
    return tuple(sum(matrix[row][column] * vector[column] for column in range(3)) for row in range(3))


def mat_t_vec(matrix, vector):
    return tuple(sum(matrix[row][column] * vector[row] for row in range(3)) for column in range(3))


def project(point, camera_transform, width: int, height: int, lens: float, sensor_width: float):
    camera_location, camera_rotation = camera_transform
    camera_point = mat_t_vec(camera_rotation, subtract(point, camera_location))
    depth = -camera_point[2]
    if depth <= 0.0:
        return None
    sensor_height = sensor_width * height / width
    u = 0.5 + lens * camera_point[0] / (depth * sensor_width)
    v_bottom = 0.5 + lens * camera_point[1] / (depth * sensor_height)
    return u * width - 0.5, (1.0 - v_bottom) * height - 0.5, depth


def dimensions(spec: dict, owner: dict) -> tuple[float, float]:
    surfaces = spec["sceneContract"]["surfaces"]
    values = surfaces["backgroundSizeWorld"] if owner["role"] == "background" else surfaces["occluderSizeWorld"]
    return float(values[0]), float(values[1])


def oracle_pixel(spec: dict, fixture: dict, x: int, y: int):
    width, height = fixture["resolution"]
    camera_spec = spec["sceneContract"]["camera"]
    lens, sensor_width = float(camera_spec["lensMm"]), float(camera_spec["sensorWidthMm"])
    sensor_height = sensor_width * height / width
    current_camera = transform(fixture["cameraByFrame"]["1"])
    previous_camera = transform(fixture["cameraByFrame"]["0"])
    u = (x + 0.5) / width
    v_bottom = 1.0 - (y + 0.5) / height
    camera_direction = ((u - 0.5) * sensor_width / lens, (v_bottom - 0.5) * sensor_height / lens, -1.0)
    world_direction = mat_vec(current_camera[1], camera_direction)
    candidates = []
    for owner_index, owner in enumerate(fixture["owners"], start=1):
        current_owner = transform(owner["transformByFrame"]["1"])
        normal = mat_vec(current_owner[1], (0.0, 0.0, 1.0))
        denominator = dot(world_direction, normal)
        if abs(denominator) < 1e-12:
            continue
        distance = dot(subtract(current_owner[0], current_camera[0]), normal) / denominator
        if distance <= 0.0:
            continue
        world_point = add(current_camera[0], scale(world_direction, distance))
        local_point = mat_t_vec(current_owner[1], subtract(world_point, current_owner[0]))
        size_x, size_y = dimensions(spec, owner)
        if abs(local_point[0]) <= size_x / 2.0 and abs(local_point[1]) <= size_y / 2.0:
            projected = project(world_point, current_camera, width, height, lens, sensor_width)
            if projected is not None:
                candidates.append((projected[2], owner_index, owner, local_point, world_point))
    if not candidates:
        return None
    current_depth, owner_index, owner, local_point, _world_point = min(candidates, key=lambda row: row[0])
    previous_owner = transform(owner["transformByFrame"]["0"])
    previous_world = add(previous_owner[0], mat_vec(previous_owner[1], local_point))
    previous_projected = project(previous_world, previous_camera, width, height, lens, sensor_width)
    if previous_projected is None:
        return None
    previous_x, previous_y, previous_depth = previous_projected
    return {
        "ownerIndex": owner_index,
        "passIndex": np.float32(owner["passIndex"]),
        "expectedVector": (previous_x - x, y - previous_y),
        "currentDepth": current_depth,
        "previousDepth": previous_depth,
    }


def taps_and_weights(qx: float, qy: float, width: int, height: int):
    x0, y0 = math.floor(qx), math.floor(qy)
    if x0 < 0 or y0 < 0 or x0 + 1 >= width or y0 + 1 >= height:
        return None
    fx, fy = qx - x0, qy - y0
    return (
        ((y0, x0), (y0, x0 + 1), (y0 + 1, x0), (y0 + 1, x0 + 1)),
        ((1.0 - fx) * (1.0 - fy), fx * (1.0 - fy), (1.0 - fx) * fy, fx * fy),
    )


def weighted(values: list[float], weights: tuple[float, ...]) -> float:
    return ((values[0] * weights[0] + values[1] * weights[1]) + values[2] * weights[2]) + values[3] * weights[3]


def neighborhood(arrays: dict, x: int, y: int, radius: int, owner: np.float32, width: int, height: int) -> bool:
    if x < radius or y < radius or x >= width - radius or y >= height - radius:
        return False
    return all(
        arrays["currentOwner"][ty, tx] == owner and arrays["currentRgba"][ty, tx, 3] > np.float32(0.999)
        for ty in range(y - radius, y + radius + 1)
        for tx in range(x - radius, x + radius + 1)
    )


def main() -> None:
    args = arguments()
    spec = json.loads(args.spec.read_text())
    if sha_file(args.spec) != SPEC_SHA256 or sha_file(Path(sys.executable)) != spec["runtime"]["python"]["sha256"] or np.__version__ != spec["runtime"]["python"]["numpy"]:
        raise RuntimeError("D12.8 Python/spec identity mismatch")
    fixture = next((row for row in spec["fixtures"] if row["id"] == args.fixture), None)
    if fixture is None or args.output_dir.exists() or args.report.exists():
        raise RuntimeError("D12.8 consumer fixture or output invalid")
    adapter = json.loads(args.adapter_report.read_text())
    body_without_hash = {key: value for key, value in adapter.items() if key != "reportHash"}
    if adapter.get("reportHash") != canonical_hash(body_without_hash) or adapter.get("fixtureId") != args.fixture or adapter.get("repeat") != args.repeat:
        raise RuntimeError("D12.8 adapter identity mismatch")
    width, height = fixture["resolution"]
    arrays = {}
    for name, (filename, channels) in INPUTS.items():
        payload = (args.input_dir / filename).read_bytes()
        if sha_bytes(payload) != adapter["arrays"][name]["sha256"]:
            raise RuntimeError(f"D12.8 input hash mismatch: {name}")
        shape = (height, width, channels) if channels > 1 else (height, width)
        arrays[name] = np.frombuffer(payload, dtype="<f4").reshape(shape).copy()
    outputs = {
        "adaptiveReconstructed": arrays["currentRgba"].copy(),
        "reason": np.zeros((height, width), dtype=np.uint8),
        "analyticOwner": np.zeros((height, width), dtype=np.uint8),
        "structuralValid": np.zeros((height, width), dtype=np.uint8),
        "radius2Interior": np.zeros((height, width), dtype=np.uint8),
        "radius3Interior": np.zeros((height, width), dtype=np.uint8),
        "adaptiveInterior": np.zeros((height, width), dtype=np.uint8),
        "adaptiveRejected": np.zeros((height, width), dtype=np.uint8),
        "riskRgb": np.zeros((height, width, 3), dtype="<f8"),
    }
    threshold = float(spec["frozenGates"]["adaptiveQuality"]["rgbMaximum"])
    for y in range(height):
        for x in range(width):
            oracle = oracle_pixel(spec, fixture, x, y)
            if oracle is None:
                outputs["reason"][y, x] = REASONS["INVALID_CURRENT_ORACLE"]
                continue
            outputs["analyticOwner"][y, x] = oracle["ownerIndex"]
            current_depth_tolerance = max(1.0, oracle["currentDepth"]) / 1024.0
            current_ok = arrays["currentOwner"][y, x] == oracle["passIndex"] and abs(float(arrays["currentDepth"][y, x]) - oracle["currentDepth"]) <= current_depth_tolerance
            if not current_ok:
                outputs["reason"][y, x] = REASONS["INVALID_CURRENT_ORACLE"]
                continue
            vector_x, vector_y = (float(value) for value in arrays["vector"][y, x])
            sample = taps_and_weights(x + vector_x, y - vector_y, width, height)
            if sample is None:
                outputs["reason"][y, x] = REASONS["INVALID_BOUNDS"]
                continue
            taps, weights = sample
            if not all(arrays["previousOwner"][ty, tx] == oracle["passIndex"] for ty, tx in taps):
                outputs["reason"][y, x] = REASONS["INVALID_OWNER"]
                continue
            if arrays["currentRgba"][y, x, 3] <= np.float32(0.999) or not all(arrays["previousRgba"][ty, tx, 3] > np.float32(0.999) for ty, tx in taps):
                outputs["reason"][y, x] = REASONS["INVALID_ALPHA"]
                continue
            sampled_depth = weighted([float(arrays["previousDepth"][ty, tx]) for ty, tx in taps], weights)
            depth_tolerance = max(1.0, oracle["previousDepth"]) / 1024.0
            if abs(sampled_depth - oracle["previousDepth"]) > depth_tolerance:
                outputs["reason"][y, x] = REASONS["INVALID_DEPTH"]
                continue
            outputs["reason"][y, x] = REASONS["VALID"]
            outputs["structuralValid"][y, x] = 1
            r2 = neighborhood(arrays, x, y, 2, oracle["passIndex"], width, height)
            r3 = neighborhood(arrays, x, y, 3, oracle["passIndex"], width, height)
            outputs["radius2Interior"][y, x] = int(r2)
            outputs["radius3Interior"][y, x] = int(r3)
            if not r2:
                continue
            reconstructed = np.empty((4,), dtype="<f4")
            for channel in range(4):
                values = [float(arrays["previousRgba"][ty, tx, channel]) for ty, tx in taps]
                reconstructed[channel] = np.float32(weighted(values, weights))
                if channel < 3:
                    center = float(arrays["currentRgba"][y, x, channel])
                    outputs["riskRgb"][y, x, channel] = sum(abs(weight) * abs(value - center) for weight, value in zip(weights, values)) + abs(float(np.spacing(reconstructed[channel])))
            if float(outputs["riskRgb"][y, x].max()) <= threshold:
                outputs["adaptiveInterior"][y, x] = 1
                outputs["adaptiveReconstructed"][y, x] = reconstructed
            else:
                outputs["adaptiveRejected"][y, x] = 1
    args.output_dir.mkdir(parents=True, exist_ok=False)
    records = {}
    for name, (filename, dtype) in OUTPUTS.items():
        payload = np.ascontiguousarray(outputs[name], dtype=dtype).tobytes()
        target = args.output_dir / filename
        target.write_bytes(payload)
        records[name] = {"uri": str(target), "sha256": sha_bytes(payload), "bytes": len(payload), "shape": list(outputs[name].shape), "dtype": "uint8" if dtype == "u1" else ("little-endian-float64" if dtype == "<f8" else "little-endian-float32")}
    body = {
        "schemaVersion": "bfs.blenderProjectiveMotionDisocclusionConsumerReport.v0.1",
        "experimentId": spec["experimentId"],
        "producer": "python",
        "fixtureId": args.fixture,
        "repeat": args.repeat,
        "pid": os.getpid(),
        "runtime": {"python": sys.version.split()[0], "pythonExecutableSha256": sha_file(Path(sys.executable)), "numpy": np.__version__},
        "adapter": {"uri": str(args.adapter_report), "sha256": sha_file(args.adapter_report), "reportHash": adapter["reportHash"]},
        "reasonCodes": REASONS,
        "projectionContract": spec["projectionOracle"],
        "structuralContract": spec["structuralValidity"],
        "candidateContract": spec["candidateContract"],
        "arrays": records,
        "operationCounts": {"consumerProcesses": 1, "pixelsVisited": width * height, "modelCalls": 0, "networkCalls": 0},
    }
    report = {**body, "reportHash": canonical_hash(body)}
    args.report.parent.mkdir(parents=True, exist_ok=True)
    args.report.write_text(json.dumps(report, indent=2, sort_keys=True, allow_nan=False) + "\n")
    print(f"BFS_B52_D128_CONSUMER_PY_OK fixture={args.fixture} repeat={args.repeat}")


if __name__ == "__main__":
    main()

File: scripts/reconstruct-b52-d12-8-motion-disocclusion.mjs
#!/usr/bin/env node
/** Scalar Node structural-validity and adaptive-risk consumer for B52-D12.8. */
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';

const SPEC_SHA256 = 'd7e7c0ee0bd7f512766188eabda9fa0dccb098a0729b26487aa38bee97d6aea6';
const INPUTS = {
  previousRgba: ['previous.rgba32', 4], currentRgba: ['current.rgba32', 4],
  previousDepth: ['previous-depth.f32', 1], currentDepth: ['current-depth.f32', 1],
  previousOwner: ['previous-owner.f32', 1], currentOwner: ['current-owner.f32', 1],
  vector: ['vector.xy32', 2], vectorNext: ['vector-next.xy32', 2],
};
const OUTPUTS = {
  adaptiveReconstructed: ['adaptive-reconstructed.rgba32', 'f32'], reason: ['reason.u8', 'u8'],
  analyticOwner: ['analytic-owner.u8', 'u8'], structuralValid: ['structural-valid.u8', 'u8'],
  radius2Interior: ['radius2-interior.u8', 'u8'], radius3Interior: ['radius3-interior.u8', 'u8'],
  adaptiveInterior: ['adaptive-interior.u8', 'u8'], adaptiveRejected: ['adaptive-rejected.u8', 'u8'],
  riskRgb: ['risk.rgb64', 'f64'],
};
const REASONS = {UNREGISTERED: 0, INVALID_CURRENT_ORACLE: 1, INVALID_BOUNDS: 2, INVALID_OWNER: 3, INVALID_ALPHA: 4, INVALID_DEPTH: 5, VALID: 6};

function shaBytes(value) { return crypto.createHash('sha256').update(value).digest('hex'); }
function shaFile(filePath) { return shaBytes(fs.readFileSync(filePath)); }
function stable(value) {
  if (Array.isArray(value)) return `[${value.map(stable).join(',')}]`;
  if (value && typeof value === 'object') return `{${Object.keys(value).sort().map(key => `${JSON.stringify(key)}:${stable(value[key])}`).join(',')}}`;
  return JSON.stringify(value);
}
function canonicalHash(value) { return shaBytes(Buffer.from(stable(value))); }
function parseArgs() {
  const result = {};
  for (let index = 2; index < process.argv.length; index += 2) {
    const key = process.argv[index];
    if (!key.startsWith('--') || index + 1 >= process.argv.length) throw new Error('invalid D12.8 arguments');
    result[key.slice(2)] = process.argv[index + 1];
  }
  for (const key of ['spec', 'fixture', 'repeat', 'input-dir', 'adapter-report', 'output-dir', 'report']) if (!(key in result)) throw new Error(`missing --${key}`);
  result.repeat = Number(result.repeat);
  if (![1, 2].includes(result.repeat)) throw new Error('invalid D12.8 repeat');
  return result;
}
function readF32(filePath, expectedCount) {
  const payload = fs.readFileSync(filePath);
  if (payload.length !== expectedCount * 4) throw new Error(`D12.8 float32 length mismatch: ${filePath}`);
  const result = new Float32Array(expectedCount);
  for (let index = 0; index < expectedCount; index++) result[index] = payload.readFloatLE(index * 4);
  return [result, payload];
}
function rotationXyz(values) {
  const [x, y, z] = values.map(Number); const cx = Math.cos(x), sx = Math.sin(x), cy = Math.cos(y), sy = Math.sin(y), cz = Math.cos(z), sz = Math.sin(z);
  return [[cz * cy, cz * sy * sx - sz * cx, cz * sy * cx + sz * sx], [sz * cy, sz * sy * sx + cz * cx, sz * sy * cx - cz * sx], [-sy, cy * sx, cy * cx]];
}
function transform(row) { return [row.location.map(Number), rotationXyz(row.rotationEuler)]; }
const add = (a, b) => [a[0] + b[0], a[1] + b[1], a[2] + b[2]];
const subtract = (a, b) => [a[0] - b[0], a[1] - b[1], a[2] - b[2]];
const scale = (a, value) => [a[0] * value, a[1] * value, a[2] * value];
const dot = (a, b) => a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
function matVec(matrix, vector) { return [0, 1, 2].map(row => matrix[row][0] * vector[0] + matrix[row][1] * vector[1] + matrix[row][2] * vector[2]); }
function matTVec(matrix, vector) { return [0, 1, 2].map(column => matrix[0][column] * vector[0] + matrix[1][column] * vector[1] + matrix[2][column] * vector[2]); }
function project(point, cameraTransform, width, height, lens, sensorWidth) {
  const cameraPoint = matTVec(cameraTransform[1], subtract(point, cameraTransform[0])); const depth = -cameraPoint[2];
  if (depth <= 0) return null;
  const sensorHeight = sensorWidth * height / width; const u = 0.5 + lens * cameraPoint[0] / (depth * sensorWidth); const vBottom = 0.5 + lens * cameraPoint[1] / (depth * sensorHeight);
  return [u * width - 0.5, (1 - vBottom) * height - 0.5, depth];
}
function dimensions(spec, owner) { return spec.sceneContract.surfaces[owner.role === 'background' ? 'backgroundSizeWorld' : 'occluderSizeWorld'].map(Number); }
function oraclePixel(spec, fixture, x, y) {
  const [width, height] = fixture.resolution; const cameraSpec = spec.sceneContract.camera; const lens = Number(cameraSpec.lensMm), sensorWidth = Number(cameraSpec.sensorWidthMm), sensorHeight = sensorWidth * height / width;
  const currentCamera = transform(fixture.cameraByFrame['1']), previousCamera = transform(fixture.cameraByFrame['0']);
  const u = (x + 0.5) / width, vBottom = 1 - (y + 0.5) / height;
  const cameraDirection = [(u - 0.5) * sensorWidth / lens, (vBottom - 0.5) * sensorHeight / lens, -1]; const worldDirection = matVec(currentCamera[1], cameraDirection); const candidates = [];
  fixture.owners.forEach((owner, zeroIndex) => {
    const currentOwner = transform(owner.transformByFrame['1']), normal = matVec(currentOwner[1], [0, 0, 1]), denominator = dot(worldDirection, normal);
    if (Math.abs(denominator) < 1e-12) return;
    const distance = dot(subtract(currentOwner[0], currentCamera[0]), normal) / denominator; if (distance <= 0) return;
    const worldPoint = add(currentCamera[0], scale(worldDirection, distance)); const localPoint = matTVec(currentOwner[1], subtract(worldPoint, currentOwner[0])); const [sizeX, sizeY] = dimensions(spec, owner);
    if (Math.abs(localPoint[0]) <= sizeX / 2 && Math.abs(localPoint[1]) <= sizeY / 2) {
      const projected = project(worldPoint, currentCamera, width, height, lens, sensorWidth); if (projected) candidates.push([projected[2], zeroIndex + 1, owner, localPoint]);
    }
  });
  if (!candidates.length) return null; candidates.sort((left, right) => left[0] - right[0]); const [currentDepth, ownerIndex, owner, localPoint] = candidates[0];
  const previousOwner = transform(owner.transformByFrame['0']); const previousWorld = add(previousOwner[0], matVec(previousOwner[1], localPoint)); const previous = project(previousWorld, previousCamera, width, height, lens, sensorWidth); if (!previous) return null;
  return {ownerIndex, passIndex: Math.fround(owner.passIndex), expectedVector: [previous[0] - x, y - previous[1]], currentDepth, previousDepth: previous[2]};
}
function tapsAndWeights(qx, qy, width, height) {
  const x0 = Math.floor(qx), y0 = Math.floor(qy); if (x0 < 0 || y0 < 0 || x0 + 1 >= width || y0 + 1 >= height) return null;
  const fx = qx - x0, fy = qy - y0; return {taps: [[y0, x0], [y0, x0 + 1], [y0 + 1, x0], [y0 + 1, x0 + 1]], weights: [(1 - fx) * (1 - fy), fx * (1 - fy), (1 - fx) * fy, fx * fy]};
}
function weighted(values, weights) { return ((values[0] * weights[0] + values[1] * weights[1]) + values[2] * weights[2]) + values[3] * weights[3]; }
function f32Ulp(value) {
  const buffer = Buffer.alloc(4); buffer.writeFloatLE(Math.fround(value), 0); let bits = buffer.readUInt32LE(0); if ((bits & 0x7fffffff) === 0) return 2 ** -149;
  bits += (bits & 0x80000000) ? -1 : 1; const next = Buffer.alloc(4); next.writeUInt32LE(bits >>> 0, 0); return Math.abs(next.readFloatLE(0) - Math.fround(value));
}
function encode(array, type) {
  if (type === 'u8') return Buffer.from(array);
  const bytes = type === 'f64' ? 8 : 4, payload = Buffer.alloc(array.length * bytes);
  for (let index = 0; index < array.length; index++) type === 'f64' ? payload.writeDoubleLE(array[index], index * 8) : payload.writeFloatLE(array[index], index * 4);
  return payload;
}
function main() {
  const args = parseArgs(), spec = JSON.parse(fs.readFileSync(args.spec, 'utf8'));
  if (shaFile(args.spec) !== SPEC_SHA256 || shaFile(process.execPath) !== spec.runtime.node.sha256 || process.version !== spec.runtime.node.version) throw new Error('D12.8 Node/spec identity mismatch');
  const fixture = spec.fixtures.find(row => row.id === args.fixture); if (!fixture || fs.existsSync(args['output-dir']) || fs.existsSync(args.report)) throw new Error('D12.8 consumer fixture or output invalid');
  const adapter = JSON.parse(fs.readFileSync(args['adapter-report'], 'utf8')), adapterBody = Object.fromEntries(Object.entries(adapter).filter(([key]) => key !== 'reportHash'));
  if (adapter.reportHash !== canonicalHash(adapterBody) || adapter.fixtureId !== args.fixture || adapter.repeat !== args.repeat) throw new Error('D12.8 adapter identity mismatch');
  const [width, height] = fixture.resolution, pixels = width * height, arrays = {};
  for (const [name, [filename, channels]] of Object.entries(INPUTS)) { const [value, payload] = readF32(path.join(args['input-dir'], filename), pixels * channels); if (shaBytes(payload) !== adapter.arrays[name].sha256) throw new Error(`D12.8 input hash mismatch: ${name}`); arrays[name] = value; }
  const outputs = {adaptiveReconstructed: new Float32Array(arrays.currentRgba), reason: new Uint8Array(pixels), analyticOwner: new Uint8Array(pixels), structuralValid: new Uint8Array(pixels), radius2Interior: new Uint8Array(pixels), radius3Interior: new Uint8Array(pixels), adaptiveInterior: new Uint8Array(pixels), adaptiveRejected: new Uint8Array(pixels), riskRgb: new Float64Array(pixels * 3)};
  const threshold = Number(spec.frozenGates.adaptiveQuality.rgbMaximum), rgba = (pixel, channel) => pixel * 4 + channel, xy = (pixel, channel) => pixel * 2 + channel;
  function neighborhood(x, y, radius, owner) { if (x < radius || y < radius || x >= width - radius || y >= height - radius) return false; for (let ty = y - radius; ty <= y + radius; ty++) for (let tx = x - radius; tx <= x + radius; tx++) { const pixel = ty * width + tx; if (arrays.currentOwner[pixel] !== owner || arrays.currentRgba[rgba(pixel, 3)] <= Math.fround(0.999)) return false; } return true; }
  for (let y = 0; y < height; y++) for (let x = 0; x < width; x++) {
    const pixel = y * width + x, oracle = oraclePixel(spec, fixture, x, y); if (!oracle) { outputs.reason[pixel] = REASONS.INVALID_CURRENT_ORACLE; continue; }
    outputs.analyticOwner[pixel] = oracle.ownerIndex;
    const currentOk = arrays.currentOwner[pixel] === oracle.passIndex && Math.abs(arrays.currentDepth[pixel] - oracle.currentDepth) <= Math.max(1, oracle.currentDepth) / 1024; if (!currentOk) { outputs.reason[pixel] = REASONS.INVALID_CURRENT_ORACLE; continue; }
    const vectorX = arrays.vector[xy(pixel, 0)], vectorY = arrays.vector[xy(pixel, 1)], sample = tapsAndWeights(x + vectorX, y - vectorY, width, height); if (!sample) { outputs.reason[pixel] = REASONS.INVALID_BOUNDS; continue; }
    const tapPixels = sample.taps.map(([ty, tx]) => ty * width + tx); if (!tapPixels.every(tap => arrays.previousOwner[tap] === oracle.passIndex)) { outputs.reason[pixel] = REASONS.INVALID_OWNER; continue; }
    if (arrays.currentRgba[rgba(pixel, 3)] <= Math.fround(0.999) || !tapPixels.every(tap => arrays.previousRgba[rgba(tap, 3)] > Math.fround(0.999))) { outputs.reason[pixel] = REASONS.INVALID_ALPHA; continue; }
    const sampledDepth = weighted(tapPixels.map(tap => arrays.previousDepth[tap]), sample.weights); if (Math.abs(sampledDepth - oracle.previousDepth) > Math.max(1, oracle.previousDepth) / 1024) { outputs.reason[pixel] = REASONS.INVALID_DEPTH; continue; }
    outputs.reason[pixel] = REASONS.VALID; outputs.structuralValid[pixel] = 1; const r2 = neighborhood(x, y, 2, oracle.passIndex), r3 = neighborhood(x, y, 3, oracle.passIndex); outputs.radius2Interior[pixel] = Number(r2); outputs.radius3Interior[pixel] = Number(r3); if (!r2) continue;
    const reconstructed = new Float32Array(4); for (let channel = 0; channel < 4; channel++) { const values = tapPixels.map(tap => arrays.previousRgba[rgba(tap, channel)]); reconstructed[channel] = Math.fround(weighted(values, sample.weights)); if (channel < 3) { const center = arrays.currentRgba[rgba(pixel, channel)]; outputs.riskRgb[pixel * 3 + channel] = sample.weights.reduce((sum, weight, index) => sum + Math.abs(weight) * Math.abs(values[index] - center), 0) + f32Ulp(reconstructed[channel]); } }
    const riskMax = Math.max(outputs.riskRgb[pixel * 3], outputs.riskRgb[pixel * 3 + 1], outputs.riskRgb[pixel * 3 + 2]); if (riskMax <= threshold) { outputs.adaptiveInterior[pixel] = 1; for (let channel = 0; channel < 4; channel++) outputs.adaptiveReconstructed[rgba(pixel, channel)] = reconstructed[channel]; } else outputs.adaptiveRejected[pixel] = 1;
  }
  fs.mkdirSync(args['output-dir'], {recursive: true}); const records = {};
  for (const [name, [filename, type]] of Object.entries(OUTPUTS)) { const payload = encode(outputs[name], type), target = path.join(args['output-dir'], filename); fs.writeFileSync(target, payload); records[name] = {uri: target, sha256: shaBytes(payload), bytes: payload.length, shape: name === 'adaptiveReconstructed' ? [height, width, 4] : name === 'riskRgb' ? [height, width, 3] : [height, width], dtype: type === 'u8' ? 'uint8' : type === 'f64' ? 'little-endian-float64' : 'little-endian-float32'}; }
  const body = {schemaVersion: 'bfs.blenderProjectiveMotionDisocclusionConsumerReport.v0.1', experimentId: spec.experimentId, producer: 'node', fixtureId: args.fixture, repeat: args.repeat, pid: process.pid, runtime: {node: process.version, nodeExecutableSha256: shaFile(process.execPath)}, adapter: {uri: args['adapter-report'], sha256: shaFile(args['adapter-report']), reportHash: adapter.reportHash}, reasonCodes: REASONS, projectionContract: spec.projectionOracle, structuralContract: spec.structuralValidity, candidateContract: spec.candidateContract, arrays: records, operationCounts: {consumerProcesses: 1, pixelsVisited: pixels, modelCalls: 0, networkCalls: 0}};
  const report = {...body, reportHash: canonicalHash(body)}; fs.mkdirSync(path.dirname(args.report), {recursive: true}); fs.writeFileSync(args.report, `${JSON.stringify(report, null, 2)}\n`); console.log(`BFS_B52_D128_CONSUMER_NODE_OK fixture=${args.fixture} repeat=${args.repeat}`);
}
main();
